converter.go

```diff
@@ -1966,200 +1966,216 @@ func arrowToRecord(ctx context.Context, record arrow.Record, pool memory.Allocat
 	ctxAlloc := compute.WithAllocator(ctx, pool)
 
 	for i, col := range record.Columns() {
-		srcColumnMeta := rowType[i]
+		fieldMetadata := rowType[i].toFieldMetadata()
 
-		// TODO: confirm that it is okay to be using higher precision logic for conversions
-		newCol := col
-		snowflakeType := getSnowflakeType(srcColumnMeta.Type)
-		switch snowflakeType {
-		case fixedType:
+		newCol, err := arrowToRecordSingleColumn(ctxAlloc, s.Field(i), col, fieldMetadata, higherPrecisionEnabled, arrowBatchesTimestampOption, pool, loc, numRows)
+		if err != nil {
+			return nil, err
+		}
+		cols = append(cols, newCol)
+		defer newCol.Release()
+	}
+	newRecord := array.NewRecord(s, cols, numRows)
+	return newRecord, nil
+}
+
+func arrowToRecordSingleColumn(ctx context.Context, field arrow.Field, col arrow.Array, fieldMetadata fieldMetadata, higherPrecisionEnabled bool, timestampOption snowflakeArrowBatchesTimestampOption, pool memory.Allocator, loc *time.Location, numRows int64) (arrow.Array, error) {
+	var err error
+	newCol := col
+	snowflakeType := getSnowflakeType(fieldMetadata.Type)
+	switch snowflakeType {
+	case fixedType:
+		if higherPrecisionEnabled {
+			// do nothing - return decimal as is
+			col.Retain()
```

Comment:
Somehow I feel like this is going to cause problems

Reply:
Can you elaborate a bit? This code changed, retaining/releasing changed as well. Now, we release at the end all columns (we can't use defer because we need more than function scope). But this column is not going to be changed, so it would be released twice (with old record and new one). So we have to bump it's counter.